File: 0239-sliding-window-maximum/solution.cpp
class Solution {
public:
    //  app:- use deque ( copy past of TECH DOSE)    TC=O(N), SC=O(K)
    
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        // Pair<int,int>  == nums[i], i
        deque< pair<int, int> > dq;
        vector<int> ans;
        
        for(int i=0; i<nums.size(); i++){
            if( !dq.empty() && dq.front().second <= i-k ){  // removing front ele if it goes out of window size
                dq.pop_front();
            }
                
            while(!dq.empty()  && dq.back().first  < nums[i]){  // maintaining elements  in DEC order
                 dq.pop_back();
            }
            
            dq.push_back( make_pair(nums[i], i));  // push current NODE
            
            if( i >= k-1){
                ans.push_back(dq.front().first);  // including maxi of current window in ans vector
            }
        }
        return ans;
    }
};

File: 0239-sliding-window-maximum/solution.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        // tech dose code  but this time i have not copy (i have learn and 100% understood, then i have coded)
        Deque<Pair> dq = new LinkedList<>();
        int n = nums.length;
        int []ans = new int[n-k+1];
        int ind = 0;
        for(int i=0; i<n; i++){
            if(dq.size() >= 1 && dq.getFirst().second <= i-k ){
                dq.pollFirst();
            }

            while(dq.size() >= 1 && dq.getLast().first < nums[i]){
                dq.pollLast();
            }

            dq.addLast(new Pair(nums[i], i));

            if(i+1 >= k){
                ans[ind++] = dq.getFirst().first;
            }
        }
        return ans;
    }
}
class Pair{
    int first = 0, second = 0;
    Pair(int f, int s){
        first = f;
        second = s;
    }
}
